### PR TITLE
[Workflow] Fix recovery storage mismatch issue

### LINKED ISSUE #17097: [Workflow] During recovery, a different storage is used
### What is the problem?

See https://github.com/ray-project/ray/pull/17065#discussion_r669300299

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

Originally, the workflow would fail if we use a storage different from the current global storage.

This happens mainly because the we  did not setup the store in `update_workflow_step_context`.

I also feel the original code is too "messy", so I cleaned them up.

## Related issue number

Closes https://github.com/ray-project/ray/issues/17097

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by suquark** [comment]
@ericl @iycheng ready for review

**Comment by iycheng** [comment]
@suquark  Could you give more details about what's the problem in the PR description?

File: python/ray/experimental/workflow/execution.py
Comment (iycheng) [inline python/ray/experimental/workflow/execution.py:71]: Maybe you missed the comment last time, bud could you update the name here?
actor -> workflow_manager

File: python/ray/experimental/workflow/execution.py
Comment (iycheng) [inline python/ray/experimental/workflow/execution.py:52]: I like the commit step api here since it put the commit code into one explicitly. Any reason why you get rid of it?

Comment (suquark) [reply python/ray/experimental/workflow/execution.py:52]: It is because the function seems too "shallow", it only has 3 lines of code. After addressing the bug, the number of arguments are more than the lines of code.

File: python/ray/experimental/workflow/execution.py
Comment (suquark) [reply python/ray/experimental/workflow/execution.py:71]: hmmm, seems this is a different one. let me fix it

File: python/ray/experimental/workflow/execution.py
Comment (suquark) [reply python/ray/experimental/workflow/execution.py:52]: I'm also fine if you decide to keep it. I think it is useful when we pass storage via context.

Comment (iycheng) [reply python/ray/experimental/workflow/execution.py:52]: Let's keep commit_step here which makes it easier to understand. Besides, there is also another place commit the step.

File: python/ray/experimental/workflow/workflow_context.py
Comment (iycheng) [inline python/ray/experimental/workflow/workflow_context.py:58]: Usually, when this happens, it means something in design is wrong.

**Comment by iycheng** [comment]
I'm OK with this PR being merged to unblock other critical things. The API to manage storage should be improved. Let's discuss it and update it in another PR.

**Review by fishbone — APPROVED** [review]
Thanks for the fixing. Please fix the comment before merging it.

File: python/ray/experimental/workflow/workflow_context.py
Comment (suquark) [reply python/ray/experimental/workflow/workflow_context.py:58]: I think we need a "workflow" abstract base class. This addresses most issues.

**Pull request merged**
